File: graph_mst/kruskal/src/KruskalDemo.java
// java
// グラフの最小全域木: クラスカル法 (Kruskal)

import java.util.*;

    
/**
 * Union-Find（素集合データ構造）を実装するクラス
 */
class DSU {
    private Map<String, String> parent;
    private Map<String, Integer> rank;
    
    /**
     * コンストラクタ
     * @param vertices 頂点のリスト
     */
    public DSU(List<String> vertices) {
        parent = new HashMap<>();
        rank = new HashMap<>();
        
        // 各頂点の親を自分自身に初期化
        for (String v : vertices) {
            parent.put(v, v);
            rank.put(v, 0);
        }
    }
    
    /**
     * 頂点iが属する集合の代表元（根）を見つける
     * @param i 頂点
     * @return 根の頂点
     */
    public String find(String i) {
        if (parent.get(i).equals(i)) {
            return i;
        }
        // パス圧縮の実装
        parent.put(i, find(parent.get(i)));
        return parent.get(i);
    }
    
    /**
     * 頂点iと頂点jを含む二つの集合を結合する
     * @param i 頂点1
     * @param j 頂点2
     * @return 結合が行われた場合はtrue、すでに同じ集合に属している場合はfalse
     */
    public boolean union(String i, String j) {
        String rootI = find(i);
        String rootJ = find(j);
        
        // 根が同じ場合は何もしない
        if (rootI.equals(rootJ)) {
            return false;
        }
        
        // ランクが小さい方の木をランクが大きい方の木の根に付ける
        if (rank.get(rootI) < rank.get(rootJ)) {
            parent.put(rootI, rootJ);
        } else if (rank.get(rootI) > rank.get(rootJ)) {
            parent.put(rootJ, rootI);
        } else {
            // ランクが同じ場合はどちらかを親とし、そのランクを増やす
            parent.put(rootJ, rootI);
            rank.put(rootI, rank.get(rootI) + 1);
        }
        return true;
    }
}

/**
 * グラフデータを管理するクラス
 */
class GraphData {
    // 隣接リストによるグラフ表現
    private Map<String, List<Pair<String, Integer>>> data;
    
    /**
     * コンストラクタ
     */
    public GraphData() {
        data = new HashMap<>();
    }
    
    /**
     * グラフの内部データを取得
     * @return グラフデータ
     */
    public Map<String, List<Pair<String, Integer>>> get() {
        return data;
    }
    
    /**
     * グラフの全頂点をリストとして取得
     * @return 頂点リスト
     */
    public List<String> getVertices() {
        return new ArrayList<>(data.keySet());
    }
    
    /**
     * グラフの全辺をリストとして取得
     * @return 辺のリスト（始点、終点、重み）のタプル
     */
    public List<Triple<String, String, Integer>> getEdges() {
        Set<Triple<String, String, Integer>> edges = new HashSet<>();
        
        for (String vertex : data.keySet()) {
            for (Pair<String, Integer> neighbor : data.get(vertex)) {
                String neighborVertex = neighbor.first;
                int weight = neighbor.second;
                
                // 辺を正規化（小さい方の頂点を最初にする）
                String[] sortedVertices = new String[]{vertex, neighborVertex};
                Arrays.sort(sortedVertices);
                
                edges.add(new Triple<>(sortedVertices[0], sortedVertices[1], weight));
            }
        }
        
        return new ArrayList<>(edges);
    }
        
    /**
     * 新しい頂点をグラフに追加
     * @param vertex 追加する頂点
     * @return 成功した場合はtrue
     */
    public boolean addVertex(String vertex) {
        if (!data.containsKey(vertex)) {
            data.put(vertex, new ArrayList<>());
        }
        return true;
    }
    
    /**
     * 両頂点間に辺を追加
     * @param vertex1 頂点1
     * @param vertex2 頂点2
     * @param weight 辺の重み
     * @return 成功した場合はtrue
     */
    public boolean addEdge(String vertex1, String vertex2, int weight) {
        // 頂点が存在しない場合は追加
        if (!data.containsKey(vertex1)) {
            addVertex(vertex1);
        }
        if (!data.containsKey(vertex2)) {
            addVertex(vertex2);
        }
        
        // vertex1 -> vertex2 の辺を追加（重み付き）
        boolean edgeExistsV1V2 = false;
        for (int i = 0; i < data.get(vertex1).size(); i++) {
            Pair<String, Integer> pair = data.get(vertex1).get(i);
            if (pair.first.equals(vertex2)) {
                // 既に存在する場合は重みを更新
                data.get(vertex1).set(i, new Pair<>(vertex2, weight));
                edgeExistsV1V2 = true;
                break;
            }
        }
        if (!edgeExistsV1V2) {
            data.get(vertex1).add(new Pair<>(vertex2, weight));
        }
        
        // vertex2 -> vertex1 の辺を追加（重み付き）
        boolean edgeExistsV2V1 = false;
        for (int i = 0; i < data.get(vertex2).size(); i++) {
            Pair<String, Integer> pair = data.get(vertex2).get(i);
            if (pair.first.equals(vertex1)) {
                // 既に存在する場合は重みを更新
                data.get(vertex2).set(i, new Pair<>(vertex1, weight));
                edgeExistsV2V1 = true;
                break;
            }
        }
        if (!edgeExistsV2V1) {
            data.get(vertex2).add(new Pair<>(vertex1, weight));
        }
        
        return true;
    }
        
    /**
     * グラフを空にする
     * @return 成功した場合はtrue
     */
    public boolean clear() {
        data.clear();
        return true;
    }
    
    /**
     * 最小全域木を取得
     * @return 最小全域木の辺のリスト
     */
    public List<Triple<String, String, Integer>> getMst() {
        // 1. 全ての辺を取得し、重みでソート
        List<Triple<String, String, Integer>> edges = getEdges();
        // 重みでソート
        edges.sort(Comparator.comparing(Triple::getThird));
        
        // 2. Union-Findデータ構造を初期化
        List<String> vertices = getVertices();
        DSU dsu = new DSU(vertices);
        
        // 3. MSTを構築
        List<Triple<String, String, Integer>> mstEdges = new ArrayList<>();
        int edgesCount = 0;
        
        // ソートされた辺を順番に調べる
        for (Triple<String, String, Integer> edge : edges) {
            String u = edge.first;
            String v = edge.second;
            int weight = edge.third;
            
            // 辺の両端点が属する集合の代表元を見つける
            String rootU = dsu.find(u);
            String rootV = dsu.find(v);
            
            // 両端点が異なる集合に属する場合、その辺をMSTに追加
            if (!rootU.equals(rootV)) {
                mstEdges.add(new Triple<>(u, v, weight));
                dsu.union(u, v);
                edgesCount++;
                
                // 頂点数から1を引いた数の辺がMSTに追加されたら終了
                if (edgesCount == vertices.size() - 1) {
                    break;
                }
            }
        }
        
        return mstEdges;
    }
}

/**
 * ペア（2つの値の組）を表すクラス
 */
static class Pair<A, B> {
    public final A first;
    public final B second;
    
    public Pair(A first, B second) {
        this.first = first;
        this.second = second;
    }
    
    @Override
    public boolean equals(Object o) {
        if (this == o) return true;
        if (o == null || getClass() != o.getClass()) return false;
        Pair<?, ?> pair = (Pair<?, ?>) o;
        return Objects.equals(first, pair.first) &&
                Objects.equals(second, pair.second);
    }
    
    @Override
    public int hashCode() {
        return Objects.hash(first, second);
    }
}

/**
 * トリプル（3つの値の組）を表すクラス
 */
static class Triple<A, B, C> {
    public final A first;
    public final B second;
    public final C third;
    
    public Triple(A first, B second, C third) {
        this.first = first;
        this.second = second;
        this.third = third;
    }
    
    public C getThird() {
        return third;
    }
    
    @Override
    public boolean equals(Object o) {
        if (this == o) return true;
        if (o == null || getClass() != o.getClass()) return false;
        Triple<?, ?, ?> triple = (Triple<?, ?, ?>) o;
        return Objects.equals(first, triple.first) &&
                Objects.equals(second, triple.second) &&
                Objects.equals(third, triple.third);
    }
    
    @Override
    public int hashCode() {
        return Objects.hash(first, second, third);
    }
}

public class KruskalDemo {
    public static void main(String[] args) {
        System.out.println("Kruskal TEST -----> start");
        GraphData graphData = new GraphData();
        
        // テストケース1
        graphData.clear();
        Object[][] inputList1 = {
            {"A", "B", 4}, {"B", "C", 3}, {"B", "D", 2}, 
            {"D", "A", 1}, {"A", "C", 2}, {"B", "D", 2}
        };
        
        for (Object[] input : inputList1) {
            graphData.addEdge((String)input[0], (String)input[1], (Integer)input[2]);
        }
        
        System.out.println("\nグラフの頂点: " + graphData.getVertices());
        System.out.println("グラフの辺 (重み付き): " + graphData.getEdges());
        
        List<Triple<String, String, Integer>> outputMst1 = graphData.getMst();
        for (Triple<String, String, Integer> edge : outputMst1) {
            System.out.println("Edge: " + edge.first + " - " + edge.second + ", Weight: " + edge.third);
        }
        
        int totalWeight1 = outputMst1.stream().mapToInt(Triple::getThird).sum();
        System.out.println("最小全域木の合計重み: " + totalWeight1);
        
        // テストケース2
        graphData.clear();
        Object[][] inputList2 = {
            {"A", "B", 4}, {"C", "D", 4}, {"E", "F", 1}, {"F", "G", 1}
        };
        
        for (Object[] input : inputList2) {
            graphData.addEdge((String)input[0], (String)input[1], (Integer)input[2]);
        }
        
        System.out.println("\nグラフの頂点: " + graphData.getVertices());
        System.out.println("グラフの辺 (重み付き): " + graphData.getEdges());
        
        List<Triple<String, String, Integer>> outputMst2 = graphData.getMst();
        for (Triple<String, String, Integer> edge : outputMst2) {
            System.out.println("Edge: " + edge.first + " - " + edge.second + ", Weight: " + edge.third);
        }
        
        int totalWeight2 = outputMst2.stream().mapToInt(Triple::getThird).sum();
        System.out.println("最小全域木の合計重み: " + totalWeight2);
        
        // テストケース3
        graphData.clear();
        Object[][] inputList3 = {
            {"A", "B", 4}, {"B", "C", 3}, {"D", "E", 5}
        };
        
        for (Object[] input : inputList3) {
            graphData.addEdge((String)input[0], (String)input[1], (Integer)input[2]);
        }
        
        System.out.println("\nグラフの頂点: " + graphData.getVertices());
        System.out.println("グラフの辺 (重み付き): " + graphData.getEdges());
        
        List<Triple<String, String, Integer>> outputMst3 = graphData.getMst();
        for (Triple<String, String, Integer> edge : outputMst3) {
            System.out.println("Edge: " + edge.first + " - " + edge.second + ", Weight: " + edge.third);
        }
        
        int totalWeight3 = outputMst3.stream().mapToInt(Triple::getThird).sum();
        System.out.println("最小全域木の合計重み: " + totalWeight3);
        
        // テストケース4 (空グラフ)
        graphData.clear();
        
        System.out.println("\nグラフの頂点: " + graphData.getVertices());
        System.out.println("グラフの辺 (重み付き): " + graphData.getEdges());
        
        List<Triple<String, String, Integer>> outputMst4 = graphData.getMst();
        for (Triple<String, String, Integer> edge : outputMst4) {
            System.out.println("Edge: " + edge.first + " - " + edge.second + ", Weight: " + edge.third);
        }
        
        int totalWeight4 = outputMst4.stream().mapToInt(Triple::getThird).sum();
        System.out.println("最小全域木の合計重み: " + totalWeight4);
        
        System.out.println("\nKruskal TEST <----- end");
    }
}

File: graph_mst/kruskal/src/KruskalDemo.kt
// Kotlin
// グラフの最小全域木: クラスカル法 (Kruskal)

class DSU(vertices: List<String>) {
    // 各頂点の親を格納します。最初は各頂点自身が親です。
    private val parent = vertices.associateWith { it }.toMutableMap()
    // ランク（木の高さまたはサイズ）を格納し、union操作を最適化します。
    private val rank = vertices.associateWith { 0 }.toMutableMap()

    // 頂点 i が属する集合の代表元（根）を見つけます。
    // パス圧縮により効率化されます。
    fun find(i: String): String {
        if (parent[i] == i) {
            return i
        }
        // パス圧縮: 探索中に encountered したノードを根に直接つなぎ直します。
        parent[i] = find(parent[i]!!)
        return parent[i]!!
    }

    // 頂点 i と 頂点 j を含む二つの集合を結合します。
    // ランクによるunionにより効率化されます。
    fun union(i: String, j: String): Boolean {
        val rootI = find(i)
        val rootJ = find(j)

        // 根が異なる場合のみ結合します（同じ集合に属している場合は何もしません）。
        if (rootI != rootJ) {
            // ランクが小さい方の木を、ランクが大きい方の木の根に付けます。
            when {
                rank[rootI]!! < rank[rootJ]!! -> parent[rootI] = rootJ
                rank[rootI]!! > rank[rootJ]!! -> parent[rootJ] = rootI
                else -> {
                    // ランクが同じ場合はどちらかを親とし、その親のランクを増やします。
                    parent[rootJ] = rootI
                    rank[rootI] = rank[rootI]!! + 1
                }
            }
            return true // 集合が結合された
        }
        return false // 既に同じ集合に属していた
    }
}

// 重みを扱えるように改変された GraphData クラス
class GraphData {
    // 隣接ノードとその辺の重みを格納します。
    // キーは頂点、値はその頂点に隣接する頂点と重みのリストです。
    private val data = mutableMapOf<String, MutableList<Pair<String, Int>>>()

    fun get(): Map<String, List<Pair<String, Int>>> {
        // グラフの内部データを取得します。
        return data
    }

    fun getVertices(): List<String> {
        // グラフの全頂点をリストとして返します。
        return data.keys.toList()
    }

    fun getEdges(): List<Triple<String, String, Int>> {
        // グラフの全辺をリストとして返します。
        // 無向グラフの場合、(u, v, weight) の形式で返します。
        // 重複を避けるためにセットを使用します。
        val edges = mutableSetOf<Triple<String, String, Int>>()
        for (vertex in data.keys) {
            for ((neighbor, weight) in data[vertex]!!) {
                // 辺を正規化してセットに追加 (小さい方の頂点を最初にするなど)
                val normalizedVertices = listOf(vertex, neighbor).sorted()
                edges.add(Triple(normalizedVertices[0], normalizedVertices[1], weight))
            }
        }
        return edges.toList()
    }

    fun addVertex(vertex: String): Boolean {
        // 新しい頂点をグラフに追加します。
        if (vertex !in data) {
            data[vertex] = mutableListOf()
            return true
        }
        // 既に存在する場合は追加しないがTrueを返す（変更なしでも成功とみなす）
        return true
    }

    fun addEdge(vertex1: String, vertex2: String, weight: Int): Boolean {
        // 両頂点間に辺を追加します。重みを指定します。
        // 頂点がグラフに存在しない場合は追加します。
        if (vertex1 !in data) {
            addVertex(vertex1)
        }
        if (vertex2 !in data) {
            addVertex(vertex2)
        }
        
        // 両方向に辺を追加する（無向グラフ）
        
        // vertex1 -> vertex2 の辺を追加（重み付き）
        var edgeExistsV1V2 = false
        for (i in data[vertex1]!!.indices) {
            val (neighbor, _) = data[vertex1]!![i]
            if (neighbor == vertex2) {
                data[vertex1]!![i] = Pair(vertex2, weight) // 既に存在する場合は重みを更新
                edgeExistsV1V2 = true
                break
            }
        }
        if (!edgeExistsV1V2) {
            data[vertex1]!!.add(Pair(vertex2, weight))
        }

        // vertex2 -> vertex1 の辺を追加（重み付き）
        var edgeExistsV2V1 = false
        for (i in data[vertex2]!!.indices) {
            val (neighbor, _) = data[vertex2]!![i]
            if (neighbor == vertex1) {
                data[vertex2]!![i] = Pair(vertex1, weight) // 既に存在する場合は重みを更新
                edgeExistsV2V1 = true
                break
            }
        }
        if (!edgeExistsV2V1) {
            data[vertex2]!!.add(Pair(vertex1, weight))
        }
        
        return true
    }

    fun clear(): Boolean {
        // グラフを空にします。
        data.clear()
        return true
    }

    fun getMst(): List<Triple<String, String, Int>> {
        // 1. 全ての辺を取得し、重みでソートします。
        // getEdges() は (u, v, weight) のリストを返します。
        val edges = getEdges()
        // 重み (タプルの3番目の要素) をキーとして辺をソート
        val sortedEdges = edges.sortedBy { it.third }

        // 2. Union-Findデータ構造を初期化します。
        // 各頂点が自身の集合に属するようにします。
        val vertices = getVertices()
        val dsu = DSU(vertices)

        // 3. MSTを構築します。
        // 結果として得られるMSTの辺を格納するリスト
        val mstEdges = mutableListOf<Triple<String, String, Int>>()
        // MSTに追加された辺の数 (頂点数-1 になればMSTが完成)
        var edgesCount = 0

        // ソートされた辺を順番に調べます。
        for ((u, v, weight) in sortedEdges) {
            // 辺 (u, v) の両端点が属する集合の代表元（根）を見つけます。
            val rootU = dsu.find(u)
            val rootV = dsu.find(v)

            // 両端点が異なる集合に属する場合、その辺をMSTに追加してもサイクルは形成されません。
            if (rootU != rootV) {
                // 辺をMSTに追加します。
                mstEdges.add(Triple(u, v, weight))
                // 辺を追加したので、両端点の集合を結合します。
                dsu.union(u, v)
                // MSTに追加した辺の数を増やします。
                edgesCount++

                // 頂点数から1を引いた数の辺がMSTに追加されたら終了です。
                // これはグラフが連結である場合に限り、完全なMSTが得られたことを意味します。
                // 連結でない場合は、全ての辺を調べ終えるまで続行し、最小全域森を得ます。
                if (edgesCount == vertices.size - 1) {
                    break
                }
            }
        }

        // MST (または最小全域森) の辺のリストを返します。
        return mstEdges
    }
}

fun main() {
    println("Kruskal TEST -----> start")
    val graphData = GraphData()

    graphData.clear()
    val inputList1 = listOf(
        Triple("A", "B", 4), Triple("B", "C", 3), Triple("B", "D", 2), 
        Triple("D", "A", 1), Triple("A", "C", 2), Triple("B", "D", 2)
    )
    for ((v1, v2, weight) in inputList1) {
        graphData.addEdge(v1, v2, weight)
    }
    println("\nグラフの頂点: ${graphData.getVertices()}")
    println("グラフの辺 (重み付き): ${graphData.getEdges()}")
    val outputMst1 = graphData.getMst()
    for ((v1, v2, weight) in outputMst1) {
        println("Edge: $v1 - $v2, Weight: $weight")
    }
    val totalWeight1 = outputMst1.sumOf { it.third }
    println("最小全域木の合計重み: $totalWeight1")

    graphData.clear()
    val inputList2 = listOf(
        Triple("A", "B", 4), Triple("C", "D", 4), 
        Triple("E", "F", 1), Triple("F", "G", 1)
    )
    for ((v1, v2, weight) in inputList2) {
        graphData.addEdge(v1, v2, weight)
    }
    println("\nグラフの頂点: ${graphData.getVertices()}")
    println("グラフの辺 (重み付き): ${graphData.getEdges()}")
    val outputMst2 = graphData.getMst()
    for ((v1, v2, weight) in outputMst2) {
        println("Edge: $v1 - $v2, Weight: $weight")
    }
    val totalWeight2 = outputMst2.sumOf { it.third }
    println("最小全域木の合計重み: $totalWeight2")

    graphData.clear()
    val inputList3 = listOf(
        Triple("A", "B", 4), Triple("B", "C", 3), Triple("D", "E", 5)
    )
    for ((v1, v2, weight) in inputList3) {
        graphData.addEdge(v1, v2, weight)
    }
    println("\nグラフの頂点: ${graphData.getVertices()}")
    println("グラフの辺 (重み付き): ${graphData.getEdges()}")
    val outputMst3 = graphData.getMst()
    for ((v1, v2, weight) in outputMst3) {
        println("Edge: $v1 - $v2, Weight: $weight")
    }
    val totalWeight3 = outputMst3.sumOf { it.third }
    println("最小全域木の合計重み: $totalWeight3")

    graphData.clear()
    val inputList4 = listOf<Triple<String, String, Int>>()
    for ((v1, v2, weight) in inputList4) {
        graphData.addEdge(v1, v2, weight)
    }
    println("\nグラフの頂点: ${graphData.getVertices()}")
    println("グラフの辺 (重み付き): ${graphData.getEdges()}")
    val outputMst4 = graphData.getMst()
    for ((v1, v2, weight) in outputMst4) {
        println("Edge: $v1 - $v2, Weight: $weight")
    }
    val totalWeight4 = outputMst4.sumBy { it.third }
    println("最小全域木の合計重み: $totalWeight4")

    println("\nKruskal TEST <----- end")
}